proxy/pkg/middleware/signed_url_auth.go

```diff
@@ -0,0 +1,242 @@
+		)
+
+		if status != 0 || account == nil {
+			return nil, fmt.Errorf("no oc-credential found for %v", claims.OcisID)
+		}
+
+		claims.OcisID = account.Id
+	}
+
+	return &claims, nil
+}
+
+func (m signedURLAuth) shouldServe(req *http.Request) bool {
+	return req.URL.Query().Get("OC-Signature") != ""
+}
+
+func (m signedURLAuth) validate(req *http.Request) (err error) {
+	query := req.URL.Query()
+
+	if ok, err := m.allRequiredParametersArePresent(query); !ok {
+		return err
+	}
+
+	if ok, err := m.requestMethodMatches(req.Method, query); !ok {
+		return err
+	}
+
+	if ok, err := m.requestMethodIsAllowed(req.Method); !ok {
+		return err
+	}
+
+	if expired, err := m.urlIsExpired(query, time.Now); expired {
+		return err
+	}
+
+	if ok, err := m.signatureIsValid(req); !ok {
+		return err
+	}
+
+	return nil
+}
+
+func (m signedURLAuth) allRequiredParametersArePresent(query url.Values) (ok bool, err error) {
+	// check if required query parameters exist in given request query parameters
+	// OC-Signature - the computed signature - server will verify the request upon this REQUIRED
+	// OC-Credential - defines the user scope (shall we use the owncloud user id here - this might leak internal data ....) REQUIRED
+	// OC-Date - defined the date the url was signed (ISO 8601 UTC) REQUIRED
+	// OC-Expires - defines the expiry interval in seconds (between 1 and 604800 = 7 days) REQUIRED
+	// TODO OC-Verb - defines for which http verb the request is valid - defaults to GET OPTIONAL
+	for _, p := range []string{
+		"OC-Signature",
+		"OC-Credential",
+		"OC-Date",
+		"OC-Expires",
+		"OC-Verb",
+	} {
+		if query.Get(p) == "" {
+			return false, fmt.Errorf("required %s parameter not found", p)
+		}
+	}
+
+	return true, nil
+}
+
+func (m signedURLAuth) requestMethodMatches(meth string, query url.Values) (ok bool, err error) {
+	// check if given url query parameter OC-Verb matches given request method
+	if !strings.EqualFold(meth, query.Get("OC-Verb")) {
+		return false, errors.New("required OC-Verb parameter did not match request method")
+	}
+
+	return true, nil
+}
+
+func (m signedURLAuth) requestMethodIsAllowed(meth string) (ok bool, err error) {
+	//  check if given request method is allowed
+	methodIsAllowed := false
+	for _, am := range m.preSignedURLConfig.AllowedHTTPMethods {
+		if strings.EqualFold(meth, am) {
+			methodIsAllowed = true
+			break
+		}
+	}
+
+	if !methodIsAllowed {
+		return false, errors.New("request method is not listed in PreSignedURLConfig AllowedHTTPMethods")
+	}
+
+	return true, nil
+}
+func (m signedURLAuth) urlIsExpired(query url.Values, now func() time.Time) (expired bool, err error) {
+	// check if url is expired by checking if given date (OC-Date) + expires in seconds (OC-Expires) is after now
+	date, err := time.Parse(time.RFC3339, query.Get("OC-Date"))
+	if err != nil {
+		return true, err
+	}
+
+	expires, err := time.ParseDuration(query.Get("OC-Expires") + "s")
+	if err != nil {
+		return true, err
+	}
+
+	date.Add(expires)
```

Comment:
Oh, this should be `date = date.Add(expires)` otherwise the `Add` won't have any effect.

Reply:
good point, i also wondered what magic is going on... its a copy from the old middleware...

i update it to: 
`now > date && now < expires`

Reply:
`expires` is just a duration not a time/date you need to add it to date